[DOCS#EV-6540]: Document host policy recommendation feature (#2658)
tigera/docs: calico-enterprise/network-policy/recommendations/hep-policy-recommendations.mdx

  -p '{"metadata":{"labels":{"projectcalico.org/spec.stagedAction":"Learn"}},"spec":{"stagedAction":"Learn"}}'
```

#### Rerun policy recommendations for an enforced policy

Once a recommendation has been enforced, the engine no longer refines
it. To have the engine generate a new recommendation for a host that
already has an enforced policy, delete the enforced GlobalNetworkPolicy:

```bash
kubectl delete globalnetworkpolicy.projectcalico.org <name>
```

On the next processing cycle, the engine creates a fresh staged
recommendation in `Learn`.

### Confirm a recommendation is matching real traffic

Use `calicoctl get` with `--show-policy-activities` to check when the
recommendation engine last evaluated each staged policy against live
flow logs. Recent timestamps mean the policy is being actively
exercised by observed traffic — useful reassurance before you enforce
a recommendation.

Query a single recommendation:

```bash
calicoctl get stagedglobalnetworkpolicy <name> --show-policy-activities
```

Or every host recommendation in the `hostendpoint-isolation` tier:

```bash
calicoctl get stagedglobalnetworkpolicies --show-policy-activities \

>   -l projectcalico.org/tier=hostendpoint-isolation

Reviewer: `calicoctl get` doesn't support `-l`/`--selector` — that's a kubectl-only flag. Running this example prints the `calicoctl` usage error rather than the table shown above.

Two workable rewrites:

- Drop `-l …` and list all staged global network policies cluster-wide:
  ```bash
  calicoctl get stagedglobalnetworkpolicies --show-policy-activities
  ```
- Or chain kubectl → calicoctl by name so the tier filter still applies:
  ```bash
  kubectl get stagedglobalnetworkpolicies.projectcalico.org \
    -l projectcalico.org/tier=hostendpoint-isolation -o name \
    | sed 's|.*/||' \
    | xargs -I{} calicoctl get stagedglobalnetworkpolicy {} --show-policy-activities
  ```

Caught while validating against `seth-hep-rec` on hashrelease `2026-04-18-v3-23-hypnotic`.

- [Tune global settings](#tune-global-settings)
- [Manage the life cycle of recommendations](#manage-the-life-cycle-of-recommendations)
- [Confirm a recommendation is matching real traffic](#confirm-a-recommendation-is-matching-real-traffic)
- [Disable host-policy recommendations](#disable-host-policy-recommendations)

### Enable host-policy recommendations

Enable host-policy recommendations by setting `hostEndpointSpec.recommendationStatus`
to `Enabled` on the [PolicyRecommendationScope](../../reference/resources/policyrecommendations.mdx)
resource. When the selector is omitted, every HostEndpoint in the cluster
is eligible.

```bash
kubectl patch policyrecommendationscope default --type=merge -p '{
  "spec": {
    "hostEndpointSpec": {
      "recommendationStatus": "Enabled"
    }
  }
}'
```

:::note

Host recommendations and namespace recommendations are configured
independently, through `hostEndpointSpec` and `namespaceSpec` respectively.
Enabling one does not enable the other.

:::

By default, generated policies are placed in the `hostendpoint-isolation`
tier. You can override the tier name with `hostEndpointSpec.tierName`,

> but the tier must already exist.

Reviewer 2: Is this true?

Author: Good catch — I was wrong. Verified in `policy-recommendation/pkg/controllers/recommendation/host_recommendation_controller.go:186`: the controller calls `MaybeCreateTier(..., rectypes.PolicyRecommendationHostTierName, ...)` with the hardcoded default, so the default `hostendpoint-isolation` tier is auto-created. A custom `hostEndpointSpec.tierName` is not auto-created. Reworded in 2404f1dc.

### View recommended policies

```bash
kubectl get stagedglobalnetworkpolicies \
  -l projectcalico.org/tier=hostendpoint-isolation
```

> One staged global network policy is generated per HostEndpoint that
> matches the selector. Recommended policies move through a `Learn`,
> `Set`, and `Ignore` life cycle tracked by the
> `projectcalico.org/spec.stagedAction` label and `spec.stagedAction` field.
> See [Manage the life cycle of recommendations](#manage-the-life-cycle-of-recommendations).

Reviewer 2: ```suggestion
Each recommended policy has one of three states you can set — `Learn`, `Set`, or `Ignore` — tracked by the `projectcalico.org/spec.stagedAction` label and `spec.stagedAction` field.
See [Manage the life cycle of recommendations](#manage-the-life-cycle-of-recommendations).```

Author: Applied verbatim in 4b2291cf.

you make here apply to **both** namespace and host policy recommendations.
If namespace recommendations are also enabled on this cluster, make sure
the new values work for both.

:::

- `stabilizationPeriod` — how long a staged recommendation must remain
  unchanged before it's considered stable and ready to enforce. The
  default (`10m`) works for hosts with frequent, representative traffic.
  For hosts whose normal traffic takes longer to exercise every path,
  extend the period so the engine has time to observe a full cycle.
- `interval` — how often the engine processes new flow logs and refines
  recommendations. The default is `2m30s`.

```bash
# Extend the stabilization period to 30 minutes.
kubectl patch policyrecommendationscope default --type=merge \
  -p '{"spec":{"stabilizationPeriod":"30m"}}'

# Refine recommendations less frequently.
kubectl patch policyrecommendationscope default --type=merge \
  -p '{"spec":{"interval":"5m"}}'
```

:::tip

Review these settings soon after enabling host-policy recommendations
and adjust them based on how quickly your hosts' traffic stabilizes.
You can always revisit them later.

:::

> Other recommendation parameters (`maxRules`, `policiesLearningCutOff`,

Reviewer 2: remove this `maxRules` and `policiesLearningCutOff` references, these are currently no-ops.

Author: Confirmed no-op across the whole calico-private repo — `MaxRules` and `PoliciesLearningCutOff` are defined only on the API struct (`api/pkg/apis/projectcalico/v3/policyrecommendationscope.go:51,57`) and generated deepcopy, with zero consumers in either the namespace or HEP engine. Removed the references here and also dropped the rows from the CRD reference in 182c0467.

and the per-spec `tierName` and `selector` fields) are also tuned
through the same resource. See
[PolicyRecommendationScope](../../reference/resources/policyrecommendations.mdx)
for the full list.

### Manage the life cycle of recommendations

Staged global network policies generated by the engine move through a
three-state life cycle tracked by the
`projectcalico.org/spec.stagedAction` label and the `spec.stagedAction`

> field:

Reviewer 2: ```suggestion
Each staged global network policy generated by the engine is in one of three states, which you control through the `projectcalico.org/spec.stagedAction` label and the `spec.stagedAction` field:
```

Author: Applied verbatim in 4b2291cf.

a recommendation.

Query a single recommendation:

```bash
calicoctl get stagedglobalnetworkpolicy <name> --show-policy-activities
```

Or list every staged global network policy — host-policy recommendations
show up in the `hostendpoint-isolation` tier and can be identified by the
`TIER` column:

```bash
calicoctl get stagedglobalnetworkpolicies --show-policy-activities
```

To scope the output directly to host-policy recommendations, use
`kubectl get -o name` to resolve names in the tier and pass each one to
`calicoctl get`:

```bash
kubectl get stagedglobalnetworkpolicies.projectcalico.org \
  -l projectcalico.org/tier=hostendpoint-isolation -o name \
  | sed 's|.*/||' \
  | xargs -r -I{} calicoctl get stagedglobalnetworkpolicy {} --show-policy-activities
```

Example output:

```
NAME                                   TIER                     LAST EVALUATED              LAST EVALUATED (ANY GEN)    LAST EVALUATED GEN
hostendpoint-isolation.host-a-vfzgh    hostendpoint-isolation   2026-04-20T14:32:01-07:00   2026-04-20T14:32:01-07:00   4
hostendpoint-isolation.host-b-k9x2p    hostendpoint-isolation   NOT YET                     2026-04-20T14:20:15-07:00   2

> hostendpoint-isolation.host-c-pq4tm    hostendpoint-isolation   NOT YET                     NEVER                       N/A

Reviewer 2: Does the recommendation get created if there hasn't yet been traffic matching the host endpoint? I think this differs from the namespaced version

Author: Verified — HEP recommendations are traffic-driven. `policy-recommendation/pkg/engine/host_engine.go:208-250` (`processHostFlows`) queries flow logs first and only calls `getRecommendation(hostname)` for hostnames with observed flows (`flowsByHost`). A host with no traffic gets no staged recommendation. The namespace engine iterates `e.filteredNamespaces.All()` regardless, which is why it differs. Rewrote the NEVER explanation in 2404f1dc.

  `NOT YET` but this column has a recent timestamp, a previous version
  of the policy was matching traffic and the current version has not
  been evaluated yet — typical right after the engine refines the
  recommendation.
- **`LAST EVALUATED GEN`** — which generation the timestamp in
  `LAST EVALUATED (ANY GEN)` refers to. Compare it to `metadata.generation`
  on the staged policy to know how recent that evaluation was.

`NOT YET` means the current generation has not yet been evaluated;
`NEVER` means no generation has ever been evaluated (likely indicating
the host is not sending traffic that reaches the policy yet).

### Disable host-policy recommendations

```bash
kubectl patch policyrecommendationscope default --type=merge \
  -p '{"spec":{"hostEndpointSpec":{"recommendationStatus":"Disabled"}}}'
```

Disabling stops the engine from generating or refining host
recommendations. Existing staged and enforced policies are not removed.

## Limitations

- Host-policy recommendations are only available on standalone clusters.
- The feature has no web console surface for enabling, accepting,
  dismissing, or promoting recommendations — all workflows are
  CLI-driven.

> - The number of rules per generated policy is capped by `maxRules` on the

Reviewer 2: I believe this is a no-op, we should remove if so

Author: Confirmed no-op — the port-cap behavior described in this bullet isn't implemented (`buildPublic` in `host_engine.go:688` has no rule/port count check and never writes to the staged policy's status). Removed the bullet in 182c0467.
